File: 68.text-justification.py
class Solution:
    def fullJustify(self, words, maxWidth):
        """
        :type words: List[str]
        :type maxWidth: int
        :rtype: List[str]
        """
        ans = []
        i = 0
        while i < len(words):
            width = len(words[i])
            line = words[i]
            j = i + 1
            while j < len(words) and width + len(words[j]) + 1 <= maxWidth:
                width = width + len(words[j]) + 1
                j += 1
            if j == i + 1 or j == len(words):
                for k in range(i + 1, j):
                    line = line + " " + words[k]
                line = line.ljust(maxWidth)
            else:
                mod = (maxWidth - width) % (j - 1 - i)
                space_count = (maxWidth - width) // (j - 1 - i) + 1
                for k in range(i + 1, i + 1 + mod):
                    line = line + " " * (space_count + 1) + words[k]
                for k in range(i + 1 + mod, j):
                    line = line + " " * (space_count) + words[k]
            ans.append(line)
            i = j;
        return ans;

File: 68.text-justification.cpp
class Solution {
public:
    vector<string> fullJustify(vector<string>& words, int maxWidth) {
        vector<string> ans;
        int i = 0;
        while (i < words.size()) {
            int width = words[i].length();
            string line = words[i];
            int j = i + 1;
            while (j < words.size() && width + words[j].length() + 1 <= maxWidth) {
                width = width + words[j].length() + 1;
                j++;
            }
            if (j == i + 1 || j == words.size()) {
                for (int k = i + 1; k < j; k++) {
                    line = line + " " + words[k];
                }
                line += string(maxWidth - line.length(), ' ');
            } else {
                int mod = (maxWidth - width) % (j - 1 - i);
                int space_count = (maxWidth - width) / (j - 1 - i) + 1;
                for (int k = i + 1; k < i + 1 + mod; k++) {
                    line = line + string(space_count + 1, ' ') + words[k];
                }
                for (int k = i + 1 + mod; k < j; k++) {
                    line = line + string(space_count, ' ') + words[k];
                }
            }
            ans.push_back(line);
            i = j;
        }
        return ans;
    }
};